nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/PutHBaseRecord.java

```diff
@@ -0,0 +1,323 @@
+                columns += addBatch(tableName, flowFiles);
+                last = flowFiles.get(flowFiles.size() - 1);
+            }
+        } catch (Exception ex) {
+            getLogger().error("Failed to put records to HBase.", ex);
+            failed = true;
+        }
+
+        if (!failed) {
+            sendProvenance(session, flowFile, columns, System.nanoTime() - start, last);
+            flowFile = session.removeAttribute(flowFile, "restart.index");
+            session.transfer(flowFile, REL_SUCCESS);
+        } else {
+            String restartIndex = Integer.toString(index - flowFiles.size());
+            flowFile = session.putAttribute(flowFile, "restart.index", restartIndex);
+            sendProvenance(session, flowFile, columns, System.nanoTime() - start, last);
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+        }
+
+
+
+        session.commit();
+    }
+
+    private void sendProvenance(ProcessSession session, FlowFile flowFile, int columns, long time, PutFlowFile pff) {
+        final String details = String.format("Put %d cells to HBase.", columns);
+        session.getProvenanceReporter().send(flowFile, getTransitUri(pff), details, time);
+    }
+
+    @Override
+    protected String getTransitUri(PutFlowFile putFlowFile) {
+        return "hbase://" + putFlowFile.getTableName();
+    }
+
+
+    @Override
+    protected PutFlowFile createPut(ProcessSession session, ProcessContext context, FlowFile flowFile) {
+        return null;
+    }
+
+    protected byte[] asBytes(String field, RecordFieldType fieldType, Record record, boolean asString, String complexFieldStrategy) {
+
+        byte[] retVal;
+
+        if (asString) {
+            retVal = clientService.toBytes(record.getAsString(field));
+        } else {
+            switch (fieldType) {
+                case BOOLEAN:
+                    retVal = clientService.toBytes(record.getAsBoolean(field));
+                    break;
+                case CHAR:
+                    retVal = clientService.toBytes(record.getAsString(field));
+                    break;
+                case DOUBLE:
+                    retVal = clientService.toBytes(record.getAsDouble(field));
+                    break;
+                case FLOAT:
+                    retVal = clientService.toBytes(record.getAsFloat(field));
+                    break;
+                case INT:
+                    retVal = clientService.toBytes(record.getAsInt(field));
+                    break;
+                case LONG:
+                    retVal = clientService.toBytes(record.getAsLong(field));
+                    break;
+                default:
+                    retVal = null;
+                    switch (complexFieldStrategy) {
+                        case FAIL_VALUE:
+                            getLogger().error("Complex value found for {}; routing to failure", new Object[]{field});
```

Comment:
It seems like FAIL and WARN both log at the appropriate level and then break, which means they both return null from this method... do we need to do something different so FAIL routes to failure and WARN just skips over?